quaddtype/numpy_quaddtype/src/scalar.c

```diff
@@ -0,0 +1,253 @@
+#include <Python.h>
+#include <sleef.h>
+#include <sleefquad.h>
+#include <stdlib.h>
+
+#define PY_ARRAY_UNIQUE_SYMBOL QuadPrecType_ARRAY_API
+#define NPY_NO_DEPRECATED_API NPY_2_0_API_VERSION
+#define NO_IMPORT_ARRAY
+
+#include "numpy/arrayobject.h"
+#include "numpy/ndarraytypes.h"
+#include "numpy/dtype_api.h"
+
+#include "scalar.h"
+#include "scalar_ops.h"
+#include "dragon4.h"
+
+QuadPrecisionObject *
+QuadPrecision_raw_new(QuadBackendType backend)
+{
+    QuadPrecisionObject *new = PyObject_New(QuadPrecisionObject, &QuadPrecision_Type);
+    if (!new)
+        return NULL;
+    new->backend = backend;
+    if (backend == BACKEND_SLEEF) {
+        new->value.sleef_value = Sleef_cast_from_doubleq1(0.0);
+    }
+    else {
+        new->value.longdouble_value = 0.0L;
+    }
+    return new;
+}
+
+QuadPrecisionObject *
+QuadPrecision_from_object(PyObject *value, QuadBackendType backend)
+{
+    QuadPrecisionObject *self = QuadPrecision_raw_new(backend);
+    if (!self)
+        return NULL;
+
+    if (PyFloat_Check(value)) {
+        double dval = PyFloat_AsDouble(value);
+        if (backend == BACKEND_SLEEF) {
+            self->value.sleef_value = Sleef_cast_from_doubleq1(dval);
+        }
+        else {
+            self->value.longdouble_value = (long double)dval;
+        }
+    }
+    else if (PyUnicode_CheckExact(value)) {
+        const char *s = PyUnicode_AsUTF8(value);
+        char *endptr = NULL;
+        if (backend == BACKEND_SLEEF) {
+            self->value.sleef_value = Sleef_strtoq(s, &endptr);
+        }
+        else {
+            self->value.longdouble_value = strtold(s, &endptr);
+        }
+        if (*endptr != '\0' || endptr == s) {
+            PyErr_SetString(PyExc_ValueError, "Unable to parse string to QuadPrecision");
+            Py_DECREF(self);
+            return NULL;
+        }
+    }
+    else if (PyLong_Check(value)) {
+        long long val = PyLong_AsLongLong(value);
+        if (val == -1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_OverflowError, "Overflow Error, value out of range");
+            Py_DECREF(self);
+            return NULL;
+        }
+        if (backend == BACKEND_SLEEF) {
+            self->value.sleef_value = Sleef_cast_from_int64q1(val);
+        }
+        else {
+            self->value.longdouble_value = (long double)val;
+        }
+    }
+    else if (Py_TYPE(value) == &QuadPrecision_Type)
+    {
+        // todo: not working for ld backend, getting garbage value not sure why?
```

Comment:
can you add a failing test for this to the test suite? I can try to poke around at fixing it...

Reply:
This is fixed now, it wasn't working because by default SLEEF backend Quad was creating for all inputs leading to garbage value.

Will push the future edits by today along with the review resolves